Configuration based TUDelft calculation

joemoorhouse opened the pull request:
Finish changes related to running a flood calculation using TUDelft data, including standard of protection, and using configuration-based vulnerability.
Partial implementation of vulnerability function uncertainty in config-based models and add to docs.

xbarra reviewed (changes requested):
Please, rebase this branch onto main before merging.

xbarra reviewed (changes requested):
I would rather this got moved to it's own pr and merged.

Also a part of a bigger conversation, but squashing it on this merge would make it difficult to track

VMarfima commented on src/physrisk/vulnerability_models/config_based_impact_curves.py:
```diff
@@ -0,0 +1,349 @@
+        return key
+
+
+class ImpactCurve:
```
Missing docstring in public class(fails Ruff rule D101)

VMarfima commented on src/physrisk/vulnerability_models/config_based_impact_curves.py:
```diff
@@ -0,0 +1,349 @@
+
+
+class ImpactCurve:
+    def __init__(
```
Missing docstring in __init__ (fails Ruff rule D107)

VMarfima commented on src/physrisk/vulnerability_models/config_based_impact_curves.py:
```diff
@@ -0,0 +1,349 @@
+        self.activation_of_points_x = activation_of_points_x
+
+    @abstractmethod
+    def get_impact(self, points_x: np.ndarray) -> np.ndarray: ...
```
Missing docstring in public method (fails Ruff rule D102)

VMarfima commented on src/physrisk/vulnerability_models/config_based_impact_curves.py:
```diff
@@ -0,0 +1,349 @@
+    def get_impact(self, points_x: np.ndarray) -> np.ndarray: ...
+
+    @staticmethod
+    def get_impact_curve(item: VulnerabilityConfigItem):
```
Missing docstring in public method (fails Ruff rule D102)

VMarfima commented on src/physrisk/vulnerability_models/config_based_impact_curves.py:
```diff
@@ -0,0 +1,349 @@
+        )
+
+
+class PiecewiseLinearImpactCurve(ImpactCurve):
```
Missing docstring in public class(fails Ruff rule D101)

VMarfima commented on src/physrisk/vulnerability_models/config_based_impact_curves.py:
```diff
@@ -0,0 +1,349 @@
+        return points_y
+
+
+class ParametricImpactCurve(ImpactCurve):
```
Missing docstring in public class(fails Ruff rule D101)

VMarfima commented on src/physrisk/vulnerability_models/config_based_impact_curves.py:
```diff
@@ -0,0 +1,349 @@
+        self.points_x = points_x
+        self.points_y = points_y
+
+    def get_impact(self, points_x: np.ndarray) -> np.ndarray:
```
Missing docstring in public method (fails Ruff rule D102)

VMarfima commented on src/physrisk/vulnerability_models/config_based_impact_curves.py:
```diff
@@ -0,0 +1,349 @@
+
+
+class ParametricImpactCurve(ImpactCurve):
+    def __init__(
```
Missing docstring in init (fails Ruff rule D107)

VMarfima commented on src/physrisk/vulnerability_models/config_based_impact_curves.py:
```diff
@@ -0,0 +1,349 @@
+                1.0 / (1.0 - (points_y[0] / self.cap_of_points_y))
+            )
+
+    def get_impact(self, points_x: np.ndarray) -> np.ndarray:
```
Missing docstring in public method (fails Ruff rule D102)

VMarfima commented on src/physrisk/vulnerability_models/config_based_vuln_model_acute.py:
```diff
@@ -0,0 +1,187 @@
+from typing import (
```
Missing docstring in public module (fails Ruff rule D100).

VMarfima commented on src/physrisk/vulnerability_models/config_based_vuln_model_acute.py:
```diff
@@ -0,0 +1,187 @@
+ureg = UnitRegistry()
+
+
+class ConfigBasedVulnerabilityModelAcute(VulnerabilityModelAcuteBase):
```
Missing docstring in public class (fails Ruff rule D101)

VMarfima commented on src/physrisk/vulnerability_models/config_based_vuln_model_acute.py:
```diff
@@ -0,0 +1,187 @@
+        impact_id: str,
+        config_items: Sequence[VulnerabilityConfigItem],
+    ):
+        """A vulnerability model that requires only specification of a set of damage/disruption curves
```
1 blank line required between summary line and description  (fails Ruff rule D205)

First line of docstring should be in imperative mood: "A vulnerability model that requires only specification of a set of damage/disruption curves" (fails Ruff rule D401)

VMarfima commented on src/physrisk/vulnerability_models/config_based_vuln_model_acute.py:
```diff
@@ -0,0 +1,187 @@
+            key for curve in self.curves for key in curve.identifier.keys()
+        )
+
+    def get_data_requests(
```
Missing docstring in public method (fails Ruff rule D102)

VMarfima commented on src/physrisk/vulnerability_models/config_based_vuln_model_acute.py:
```diff
@@ -0,0 +1,187 @@
+                f"sop not available for hazard {hazard_type.__name__}"
+            )
+
+    def get_distributions(
```
Missing docstring in public method (fails Ruff rule D102)

VMarfima commented on src/physrisk/vulnerability_models/config_based_vuln_model_param.py:
```diff
@@ -0,0 +1,159 @@
+from typing import (
```
Missing docstring in public module (fails Ruff rule D100).

VMarfima commented on src/physrisk/vulnerability_models/config_based_vuln_model_param.py:
```diff
@@ -0,0 +1,159 @@
+)
+
+
+class ConfigBasedVulnerabilityModel(VulnerabilityModelBase):
```
Missing docstring in public class (fails Ruff rule D101)

VMarfima commented on src/physrisk/vulnerability_models/config_based_vuln_model_param.py:
```diff
@@ -0,0 +1,159 @@
+        impact_id: str,
+        config_items: Sequence[VulnerabilityConfigItem],
+    ):
+        """A vulnerability model that requires only specification of a set of damage/disruption curves
```
1 blank line required between summary line and description (fails Ruff rule D205)
First line of docstring should be in imperative mood: "A vulnerability model that requires only specification of a set of damage/disruption curves" (fails Ruff rule D401)

VMarfima commented on src/physrisk/vulnerability_models/config_based_vuln_model_param.py:
```diff
@@ -0,0 +1,159 @@
+        ):
+            self.thresholds = np.array(range(25, 60, 5))
+
+    def get_data_requests(
```
Missing docstring in public method (fails Ruff rule D102)

VMarfima commented on src/physrisk/vulnerability_models/config_based_vuln_model_param.py:
```diff
@@ -0,0 +1,159 @@
+            indicator_id=self.indicator_id,
+        )
+
+    def get_impact(
```
Missing docstring in public method (fails Ruff rule D102)

VMarfima commented on tests/data/hazard_model_store_test.py:
```diff
@@ -68,9 +68,9 @@ def _crs_shape_transform_global(
         width: int = 43200,
         height: int = 21600,
-        return_periods: Union[List[float], npt.NDArray] = [0.0],
+        index_values: Union[List[float], npt.NDArray, List[str]] = [0.0],
```
Do not use mutable data structures for argument defaults (fails Ruff rule B006).

VMarfima commented on tests/data/hazard_model_store_test.py:
```diff
@@ -118,10 +118,10 @@ def _crs_shape_transform(
         width: int,
         height: int,
-        return_periods: Union[List[float], npt.NDArray] = [0.0],
+        index_values: Union[List[float], npt.NDArray, List[str]] = [0.0],
```
Do not use mutable data structures for argument defaults (fails Ruff rule B006).

VMarfima commented on docs/user-guide/vulnerability_uncertainty.ipynb:
```diff
@@ -0,0 +1,2184 @@
+    "    fig.add_scatter(x=vuln.cdf[i, :], y=vuln.impact, name=f\"Flood depth {hazard_indicator[i]}m\")\n",
+    "# also interpolate for a depth of 0.6m and display\n",
+    "interp = vuln.interpolate_cdfs(np.array([0.6]))\n",
+    "fig.add_scatter(x=interp[0, :], y=vuln.impact, name=f\"Interpolated depth 0.6m\")\n",
```
f-string without any placeholders (fails Ruff rule F541).

joemoorhouse commented:
Hi @xbarra and @VMarfima,
Thanks for comments; as suggested I pulled out the PDM lock update into a separate PR and rebased. I took the change to add the PDM matrix CI at the same time.

A couple of changes that hit a lot of files:
1) Change of Type hint in vulnerability models from Iterable[HazardDataResponse] to Sequence[HazardDataResponse]. Probably should have been like this in the first place (not really intent to allow for generators here and length should be known), but had a case with the config-based models where it was desirable to see length of responses hence change in this PR. I also addressed some other Type hints, generally preferring Sequence over List.
Hopefully does not affect too much and straight forward to change; no good time to make such changes.
2) Starting to address some of the ruff excluded items I removed exclusion for unused imports.

Added some more doc strings, but still more to do there.

Thanks,
Joe
